### DISCUSSION TITLE
Asset checks: Try them out

### DISCUSSION THREAD

**Discussion by johannkm** [dagster-io/dagster#16266] [General]
> [!IMPORTANT]  
> Dagster asset checks are currently an experimental feature, and the API may change in future releases. During this period, checks do not consume Dagster Cloud Credits.

---

We've recently released a new `@asset_check` API as described in [RFC: Asset checks](https://github.com/dagster-io/dagster/discussions/15880). It includes:

- A new [`@asset_check` decorator](https://docs.dagster.io/_apidocs/asset-checks#dagster.asset_check) for defining checks that run in their own Op
- A new [`check_specs` argument](https://docs.dagster.io/_apidocs/assets#dagster.asset) on `@asset` for defining checks that execute in the same Op as the materialization
- An integration with DBT Tests: https://github.com/dagster-io/dagster/discussions/16527
- UI:
<img width="730" alt="265731326-24e7b939-e04c-4f5c-bcf3-344313fb1bba" src="https://github.com/dagster-io/dagster/assets/22018973/2d0b9f11-1299-4e9a-9954-8037d0173ab5">

![screenshot_2023-08-28_at_7 23 00_pm_720](https://github.com/dagster-io/dagster/assets/22018973/acf2cd10-6c69-4dca-b2b3-53d9b486b845)

We'd love to get your feedback after trying these out! A basic example to get started:

```python
from dagster import asset, asset_check, AssetCheckResult, Definitions

@asset
def my_asset():
  pass

@asset_check(asset=my_asset, description="A very useful check")
def always_pass_check():
  return AssetCheckResult(passed=True)

defs = Definitions(assets=[my_asset], asset_checks=[always_pass_check])
```

This check will run whenever you materialize `my_asset`.

See the [docs](https://docs.dagster.io/concepts/assets/asset-checks) for more.

---

Share feedback here, or in the [#data-quality-asset-checks](https://dagster.slack.com/archives/C05N0ELQA93) Slack channel.

**Comment by abi-mutinex** [comment]
Great work here team! A quick question - I know there were plans to integrate dbt tests to this. Any idea when they would come to life?

**Reply by johannkm** [reply]
We plan to ship dbt support in the release this Thursday

**Reply by rexledesma** [reply]
Hi @abi-mutinex, they are now released in `dagster-dbt==0.20.13`. Check out the discussion for more details: https://github.com/dagster-io/dagster/discussions/16527.

**Comment by dduong1603** [comment]
A couple of feedback/observations when trying out the asset checks:
- Not sure how to feel about the current UI color when the asset is materialized successfully but some checks fail 😅 I think I would still prefer the Materialized status to be green because at a quick glance I thought the asset failed to materialize.
- Similarly, I feel like there's a need for additional run statuses to distinguish an actual failed run that fails both materialization and checks, and a run that just fails some checks. Right now they all get grouped into Failure status.
- I still prefer the check bubble to be separated in one of the mocked UI. It looks cleaner for me 😅 
- I get this error when running some dbt tests (the test passed...I guess it just returned something that the check metadata didn't expect):
```sh
dagster._core.errors.DagsterInvariantViolationError: Received unexpected AssetCheckResult. It targets asset 'elementary/filtered_information_schema_columns' which is not targeted by any of the checks currently

Stack Trace:                                                                                                                                                                                                       
   File "/usr/local/lib/python3.11/site-packages/dagster/_core/execution/plan/execute_plan.py", line 273, in dagster_event_sequence_for_step                                                                        
     for step_event in check.generator(step_events):                                                                                                                                                                
   File "/usr/local/lib/python3.11/site-packages/dagster/_core/execution/plan/execute_step.py", line 475, in core_dagster_event_sequence_for_step                                                                   
     for user_event in _step_output_error_checked_user_event_sequence(                                                                                                                                              
   File "/usr/local/lib/python3.11/site-packages/dagster/_core/execution/plan/execute_step.py", line 158, in _step_output_error_checked_user_event_sequence                                                         
     for user_event in user_event_sequence:                                                                                                                                                                         
   File "/usr/local/lib/python3.11/site-packages/dagster/_core/execution/plan/execute_step.py", line 94, in _process_asset_results_to_events                                                                        
     yield from _process_user_event(step_context, user_event)                                                                                                                                                       
   File "/usr/local/lib/python3.11/site-packages/dagster/_core/execution/plan/execute_step.py", line 129, in _process_user_event                                                                                    
     asset_check_evaluation = user_event.to_asset_check_evaluation(step_context)                                                                                                                                    
                              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^                                                                                                                                    
   File "/usr/local/lib/python3.11/site-packages/dagster/_core/definitions/asset_check_result.py", line 88, in to_asset_check_evaluation                                                                            
     raise DagsterInvariantViolationError(
{"data": {"execution_time": 2.1847, "index": 6, "name": "elementary_schema_changes_dim__dates_", "node_info": {"materialized": "test", "meta": {"dagster": {"asset_check": true}}, "node_finished_at": "2023-10-02 
 Elementary: Started running data monitors on: WAREHOUSE.ANALYTICS_DUKE.DIM__DATES quarter_end                                                                                                                      
 Exception ignored in: <generator object DbtCliInvocation.stream_raw_events at 0x7fd4b0410f40>                                                                                                                      
 Traceback (most recent call last):                                                                                                                                                                                 
   File "/usr/local/lib/python3.11/site-packages/dagster_dbt/core/resources_v2.py", line 316, in stream                                                                                                             
     yield from event.to_default_asset_events(                                                                                                                                                                      
 RuntimeError: generator ignored GeneratorExit
```

**Reply by johannkm** [reply]
Thanks for the feedback!

> I would still prefer the Materialized status to be green because at a quick glance I thought the asset failed to materialize

If you set the check to `WARN` severity, the asset won't change colors. Does that help?

> Similarly, I feel like there's a need for additional run statuses to distinguish an actual failed run that fails both materialization and checks, and a run that just fails some checks. Right now they all get grouped into Failure status.

Could you clarify this? My expectation is that if a check emits `AssetCheckResult(success=False)`, the run will still succeed

**Reply by johannkm** [reply]
I expect the dbt error will be fixed in dagster 1.5 this week. Currently we raise that error on relationship tests and singular tests with multiple multiple models

**Reply by dduong1603** [reply]
> Could you clarify this? My expectation is that if a check emits `AssetCheckResult(success=False)`, the run will still succeed

Sorry I was referring to the UI 😄 
<img width="1498" alt="Screenshot 2023-10-02 at 3 16 55 PM" src="https://github.com/dagster-io/dagster/assets/29114884/ce78b711-ae8a-4c0a-84b6-5dcf75ad9869">

**Reply by johannkm** [reply]
Is your check raising an exception? Normally the step would succeed even if the check emits `AssetCheckResult(success=False)`

**Reply by geoHeil** [reply]
see https://dagster.slack.com/archives/C04CW71AGBW/p1696322718554279 I have a similar topic DBT gives 1 of 2 FAIL 1 not_null_my_first_dbt_model_id ................................... [FAIL 1 in 0.20s] for the tests and dagster totally stops the run in red.

**Reply by dduong1603** [reply]
Yeah even after configuring the severity to `warn` it still raised DagsterInvariantViolationError. Upgrading to dagster 1.5 still raises the same error, but the message is different
1.5:
```
dagster._core.errors.DagsterInvariantViolationError: Expected to update one row for asset check evaluation, but updated 0.
```
previously with 1.4.17:
```
dagster._core.errors.DagsterInvariantViolationError: Received unexpected AssetCheckResult. It targets asset 'elementary/filtered_information_schema_columns' which is not targeted by any of the checks currently
```
I'm using dbt 1.6.5 and the test is just a boolean type check from `dbt-expectations`

**Reply by rexledesma** [reply]
@dduong1603 What's happening is that by default, dbt is using an [eager way](https://docs.getdbt.com/reference/node-selection/test-selection-examples?indirect-selection-mode=eager) of selecting dbt tests when using `dbt test` or `dbt build`. This means that a test will run when any of its parent models are selected. 

With Dagster's asset checks, we only expect a dbt test to be run only if it is selected explicitly, or if all of its parents were selected for materialization. You can try enabling this capability locally by setting `DBT_INDIRECT_SELECTION=cautious`. Let me know if this solves your current issue for you.

In https://github.com/dagster-io/dagster/pull/16801, we fix this by turning off this test selection indirection: we set `DBT_INDIRECT_SELECTION=empty` and explicitly pass in the exact tests that should be run.

**Reply by dduong1603** [reply]
@rexledesma hmm I set `DBT_INDIRECT_SELECTION=cautious` in my user code deployment but still see the same issue. I didn't select any parent model though, and when running locally everything was ok. The failure can only be seen on the k8s deployment

**Reply by rexledesma** [reply]
@dduong1603 Glad it's running locally now. The environment variable might not be getting threaded properly in your deployment. Try setting it directly in code `os.environ["DBT_INDIRECT_SELECTION"] = "cautious"` or wait for `dagster-dbt==0.21.1` where we we'll land https://github.com/dagster-io/dagster/pull/16801 and you won't have to configure this.

**Reply by dduong1603** [reply]
Ah nice setting the variable directly in code does work!

**Reply by josh-acuitymd** [reply]
Hi @rexledesma !! Is there a way to override the default behavior of setting `DBT_INDIRECT_SELECTION=empty`? Based on [this](https://github.com/dagster-io/dagster/blob/08ae9579b3da299dc59dbacc26265dc113afc899/python_modules/libraries/dagster-dbt/dagster_dbt/core/resources_v2.py#L894-L902) it looks like it might ignore any assignments I make whether that's `eager`, `cautious`, or whatever.

**Reply by rexledesma** [reply]
See https://github.com/dagster-io/dagster/discussions/18891#discussioncomment-7949307 for an explanation of this behavior. When you enable asset checks, we require `DBT_INDIRECT_SELECTION=empty` since Dagster takes control of the exact dbt tests that are run in an execution.

**Comment by dridk** [comment]
Hi,

Thank you for this cool feature.

I recently tried mage.ai, which I find more intuitive ( more python way)  than Dagster (POO java way) in several aspects. 
For data validation, for example, mage.ai uses assertion like any other python test.

https://docs.mage.ai/development/data-validation

```
@test 
def test_number_of_rows(df) -> None:         
assert len(df.index) >= 2, 'The output has more than 1 row.'

assert df.isnull().sum() == 0 , "null values"
```



With dagster, you have to make one function per test.  Or use the yield operator in the asset. 

Why not use something like this: 

```
@asset_check 
def test1(df):
      assert len(df) > 0
      assert df.isnull().sum() = 0 

@asset_check
def test2(df): 
    ....

**Reply by johannkm** [reply]
Thanks for the feedback! The concise syntax is nice.

We have a couple concerns about using `assert`:

- asserts can be disabled at runtime with `python -O` or envvar `PYTHONOPTIMIZE`
- `AssetCheckResult` can be used inside `@asset_check`, or inside an `@asset` with multiple checks so you need to disambiguate which check you're evaluating. `assert` would only work with the former
- what if a 3rd party library you invoke calls `assert`? Should that get the overloaded behavior?

For your test1 example, how would you feel about breaking it in to separate checks so the UI can display which check failed?

```python
@asset_check(asset=df)
def check_not_empty(df):
  return AssetCheckResult(passed=len(df)==0)

@asset_check(asset=df)
def check_no_nulls(df):
  num_nulls = df.isnull().sum()
  return AssetCheckResult(passed=num_nulls>0, metadata={"number of nulls": num_nulls)
```

There's certainly more we can do to make the syntax friendly, so we're happy to get more suggestions

**Reply by dridk** [reply]
Ok ! It makes sense , particulary the point 3 !  I think it is fine in this way

**Comment by rob-apella** [comment]
can an asset check depend on multiple assets? e.g., in DBT, a singular test may join together two DBT models to perform a data test. this test would need to run after both models have been materialized.

edit: found this https://github.com/dagster-io/dagster/issues/16725

**Reply by rexledesma** [reply]
We currently only support [dbt generic tests](https://docs.getdbt.com/guides/best-practices/writing-custom-generic-tests) as Dagster asset checks. 

We're tracking support for singular tests (and generally, relationship tests that depend on multiple models) here: https://github.com/dagster-io/dagster/issues/17628

**Reply by rob-apella** [reply]
awesome thanks @rexledesma

**Reply by johannkm** [reply]
more specific tracking issue: https://github.com/dagster-io/dagster/issues/18440

**Comment by sisidra** [comment]
We want to define asset checks on source/external assets. That seems to be partially supported.
Issues:
* Dagster Web buttons "Execute" and "Execute All" are disabled
* Creating a dedicated job per check is not reasonable for large amount of assets * checks.
* External asset lack IO Manager properties (although the docs state that the goal is to replace Source Assets with External Assets) - this is more related to external assets as it is hard to integrate external assets into asset graph.

**Reply by johannkm** [reply]
Would you mind filing an issue for `Dagster Web buttons "Execute" and "Execute All" are disabled`?

**Reply by smackesey** [reply]
>External asset lack IO Manager properties (although the docs state that the goal is to replace Source Assets with External Assets) - this is more related to external assets as it is hard to integrate external assets into asset graph.

External asset APIs are under heavy development right now, we will definitely support IO managers on them.
